pilot/helpers/agents/CodeMonkey.py

```diff
@@ -1,38 +1,286 @@
-from const.function_calls import GET_FILES, IMPLEMENT_CHANGES
+import os.path
```

Comment:
This is a complete rewrite, pay no attention to the diff, it's best viewed as a completely new file.

Here's the core logic:

1. If we know which file we need to work on (as a result of us being called via `modify_file` in the project task breakdown, great). If not (which is the case when we're called from debugging), we need to figure out the file name from the code change description (ie. the instructions that come out of the debugging steps).
2. Next, we tell the LLM to identify all parts of the file that need changing, and for each part, provide a block of lines in the original, and the new block of lines that should replace the original.
3. There's a bunch of nuance there, for example LLM might not indent the blocks although they're indented in the original file so we need to compensante for that.
4. We try to find all possible matches for each code block. If there's exactly one match, great, we can do the replacement. If there are zero matches, we report an error. If there's more than one match, we report an error. We need to be careful with things like `});` of which there can be multiple, but indented differently, and we need to find all possible matches with all possible indentations, otherwise the replacement will go to an incorrect location.
5. If there was an error, we send a new message to the same conversation, saying there's an error and please retry. We do that up to 5 times (so 6 attempts in total to get it right). In most cases the LLM gets it after 2 or 3 retries.
6. If it's a really complex situation and LLM keeps getting it wrong, as a fallback we ask the LLM to just produce the output file (ie. fall back to the old behaviour but without the OLD CODE trick). Sometimes if there are a bunch of large changes, it's really easier to just output the entire file.